Replace DidFindString with a completion handler

charliewolfe opened the pull request:
#### 1a70ac5097111a3f194ee589208b5ab79c66c47d
<pre>
Replace DidFindString with a completion handler
<a href="https://bugs.webkit.org/show_bug.cgi?id=268843">https://bugs.webkit.org/show_bug.cgi?id=268843</a>
<a href="https://rdar.apple.com/122405043">rdar://122405043</a>

Reviewed by Alex Christensen.

This change moves the args passed to DidFindString to the completion handler.

When image analysis is enabled, two `DidFindString` messages are expected for one `findString:` call,
where the second message may be delayed. So, I needed to create a `FindStringInImages` message to
aggregate the results of the image search.

`DidFailToFindString` is still used by `FindStringMatches`, so I couldn’t remove that message yet.

* Source/WebKit/UIProcess/FindStringCallbackAggregator.cpp:
(WebKit::FindStringCallbackAggregator::~FindStringCallbackAggregator):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::findString):
(WebKit::WebPageProxy::didFindString): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::findStringInImages):
(WebKit::FindController::findString):
(WebKit::FindController::findStringMatches):
(WebKit::FindController::findRectsForStringMatches):
* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::findString):
(WebKit::WebPage::findStringInImages):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/274325@main">https://commits.webkit.org/274325@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/222868044376f30917f090ff4a90b167cc3224ff

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38717 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17648 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41062 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41248 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20458 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14994 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39290 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14867 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33633 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12910 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12882 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42524 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35164 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34892 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/38719 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/13527 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11188 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36926 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15139 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/13985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5048 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14620 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -208,21 +205,21 @@ void FindController::updateFindUIAfterPageScroll(bool found, const String& strin
+            matchRects = RenderObject::absoluteTextRects(*range);
+            m_findMatches.append(*range);
         }
+        completionHandler(idOfFrameContainingString, matchRects, matchCount, m_foundStringMatchIndex, didWrap == DidWrap::Yes);
```
This is an unnecessary Vector copy.  Make the parameter a Vector<IntRect>&& instead of Vector<IntRect> and you'll avoid these.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -136,38 +136,35 @@ static RefPtr<LocalFrame> frameWithSelection(Page* page)
-
-        m_webPage->send(Messages::WebPageProxy::DidFailToFindString(string));
+        if (originator != FindUIOriginator::FindString)
+            m_webPage->send(Messages::WebPageProxy::DidFailToFindString(string));
```
Can't we make this a completionHandler(std::nullopt, ...) or something?  It isn't great for there to be a completion handler that takes a result and another conditional message that also sends the results.

charliewolfe replied:
This PR makes changes so `FindString()` uses the completion handler instead of `DidFindString()` and `DidFailToFindString()`. There is also `FindStringMatches()` (for some reason) that still relies on this IPC message being sent.

charliewolfe replied:
There would need to be more changes to make `FindStringMatches()` use a completion handler. I'm fine with doing that in this PR, but it seemed like maybe I shouldn't?

achristensen07 replied:
Separate PR is fine but it really should be done.  This code is crazy without that.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -236,12 +230,24 @@ void FindController::updateFindUIAfterPageScroll(bool found, const String& strin
+void FindController::findStringInImages(const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, CompletionHandler<void(std::optional<FrameIdentifier>, Vector<IntRect>&&, uint32_t, int32_t, bool)>&& completionHandler)
+{
+    m_webPage->corePage()->analyzeImagesForFindInPage([weakPage = WeakPtr { m_webPage }, string, options, maxMatchCount, completionHandler = WTFMove(completionHandler)]() mutable {
+        if (weakPage)
```
else completionHandler({ }, { }, { }, { }, { });

danae404 commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -236,12 +230,24 @@ void FindController::updateFindUIAfterPageScroll(bool found, const String& strin
 
-void FindController::findString(const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, TriggerImageAnalysis canTriggerImageAnalysis, CompletionHandler<void(std::optional<FrameIdentifier>, bool)>&& completionHandler)
+#if ENABLE(IMAGE_ANALYSIS)
+void FindController::findStringInImages(const String& string, OptionSet<FindOptions> options, unsigned maxMatchCount, CompletionHandler<void(std::optional<FrameIdentifier>, Vector<IntRect>&&, uint32_t, int32_t, bool)>&& completionHandler)
```
`findStringInImages` makes me think that it only finds matches in images. I think a better name for this function would indicate that it finds string matches in text and in images: maybe `findStringIncludingImages`

danae404 commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -299,22 +305,11 @@ void FindController::findString(const String& string, OptionSet<FindOptions> opt
                 m_foundStringMatchIndex++;
         }
     }
-#if ENABLE(IMAGE_ANALYSIS)
-    if (canTriggerImageAnalysis == TriggerImageAnalysis::Yes) {
-        m_webPage->corePage()->analyzeImagesForFindInPage([weakPage = WeakPtr { m_webPage }, string, options, maxMatchCount] {
-            if (weakPage)
-                weakPage->findController().findString(string, options, maxMatchCount, TriggerImageAnalysis::No); 
-        });
-    }
-#endif
 
```
Technically this regresses the internal feature `imageAnalysisDuringFindInPage` because now you have to wait for image analysis to complete before getting any text matches, but I think this is okay because the increased latency will only be noticeable on pages with a lot of images, and the feature is off by default with no plans to turn it on.

charliewolfe committed:
Replace DidFindString with a completion handler
https://bugs.webkit.org/show_bug.cgi?id=268843
rdar://122405043

Reviewed by Alex Christensen.

This change moves the args passed to DidFindString to the completion handler.

When image analysis is enabled, two `DidFindString` messages are expected for one `findString:` call,
where the second message may be delayed. So, I needed to create a `FindStringInImages` message to
aggregate the results of the image search.

`DidFailToFindString` is still used by `FindStringMatches`, so I couldn’t remove that message yet.

* Source/WebKit/UIProcess/FindStringCallbackAggregator.cpp:
(WebKit::FindStringCallbackAggregator::~FindStringCallbackAggregator):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::findString):
(WebKit::WebPageProxy::didFindString): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::updateFindUIAfterPageScroll):
(WebKit::FindController::findStringInImages):
(WebKit::FindController::findString):
(WebKit::FindController::findStringMatches):
(WebKit::FindController::findRectsForStringMatches):
* Source/WebKit/WebProcess/WebPage/FindController.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::findString):
(WebKit::WebPage::findStringInImages):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/274325@main
